stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -494,16 +453,31 @@ SWIFT_CC(swift)
 static void swift_task_cancelImpl(AsyncTask *task) {
   SWIFT_TASK_DEBUG_LOG("cancel task = %p", task);
 
-  // withStatusRecordLock has some special behavior for
-  // LockContext::Cancellation; the function only gets called
-  // when they don't apply.
-  withStatusRecordLock(task, LockContext::Cancellation,
-                       [&](ActiveTaskStatus &status) {
-    assert(status.isCancelled());
+  auto oldStatus = task->_private().Status.load(std::memory_order_relaxed);
+  auto newStatus = oldStatus;
+  while (true) {
+    if (oldStatus.isCancelled()) {
+      return;
+    }
+
+    // Set cancelled bit even if oldStatus.isStatusRecordLocked()
+    newStatus = oldStatus.withCancelled();
 
-    // Carry out the cancellation operations associated with all
-    // the active records.
-    for (auto cur: status.records()) {
+    if (task->_private().Status.compare_exchange_weak(oldStatus, newStatus,
+            /*success*/ std::memory_order_relaxed,
+            /*failure*/ std::memory_order_relaxed)) {
+      break;
```

Comment:
For the reasons you mention below, you can return if there aren't any records on `oldStatus`.

We've lost an optimization to not modify the atomic multiple times when we do need to install the lock, but I guess that's not particularly important.

Reply:
Good point will add a short-circuit.

You're right that we are writing to the same atomic multiple times ...but... I do hope to be able to make a future change which allows us to avoid a full double-wide CAS when we really only care about modifying the stuff in the top word like in this loop. We should only need the double-wide CAS when we are taking the status record lock.  That should ease some register pressure at least.